ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1721,6 +1754,482 @@ void shutdown() throws InterruptedException {
+            logger.trace(
+                "Remote node: {} Thread name: {} RemoteReplicaGroup {} Create GetRequest, correlation id {}, state {}",
+                dataNodeId, threadName, id, requestInfo.getRequest().getCorrelationId(), state);
+            state = ReplicaGroupState.GET_REQUEST_SENT;
+            return requestInfo;
+          }
+        }
+        logger.trace("Remote node: {} Thread name: {} RemoteReplicaGroup {} No GetRequest needed, set state to Done",
+            dataNodeId, threadName, id);
+        state = ReplicaGroupState.DONE;
+      }
+      return null;
+    }
+
+    /**
+     * Handle the {@link ReplicaMetadataResponse} returned in {@link ResponseInfo} by nonblocking network client.
+     * @param responseInfo The {@link ResponseInfo} that carries a {@link ReplicaMetadataResponse}.
+     */
+    public void handleReplicaMetadataResponse(ResponseInfo responseInfo) {
+      if (state != ReplicaGroupState.REPLICA_METADATA_REQUEST_SENT) {
+        logger.error("Remote node: {} Thread name: {} ReplicaMetadataResponse comes back after wrong state {}",
+            remoteDataNode, threadName, state);
+        return;
+      }
+
+      NetworkClientErrorCode networkClientErrorCode = responseInfo.getError();
+      logger.trace(
+          "Remote node: {} Thread name: {} RemoteReplicaGroup {} ReplicaMetadataResponse come back for correlation id {}, NetworkClientError {}",
+          dataNodeId, threadName, id, responseInfo.getRequestInfo().getRequest().getCorrelationId(),
+          networkClientErrorCode);
+      if (networkClientErrorCode == null) {
+        replicationMetrics.updateMetadataRequestTime(time.milliseconds() - replicaMetadataRequestStartTimeMs,
+            replicatingFromRemoteColo, replicatingOverSsl, datacenterName);
+        try {
+          // Deserialize the request from the given ResponseInfo
+          DataInputStream dis = new NettyByteBufDataInputStream(responseInfo.content());
+          ReplicaMetadataResponse response = ReplicaMetadataResponse.readFrom(dis, findTokenHelper, clusterMap);
+          exchangeMetadataResponseList =
+              ReplicaThread.this.handleReplicaMetadataResponse(response, remoteReplicaInfos, remoteDataNode);
+          state = ReplicaGroupState.REPLICA_METADATA_RESPONSE_HANDLED;
+          logger.trace(
+              "Remote node: {} Thread name: {} RemoteReplicaGroup {} Handled ReplicaMetadataResponse for correlation id {}, set state to {}",
+              dataNodeId, threadName, id, responseInfo.getRequestInfo().getRequest().getCorrelationId(), state);
+          long exchangeMetadataTimeInMs = time.milliseconds() - exchangeMetadataStartTimeInMs;
+          replicationMetrics.updateExchangeMetadataTime(exchangeMetadataTimeInMs, replicatingFromRemoteColo,
+              replicatingOverSsl, datacenterName);
+        } catch (Exception e) {
+          setException(e, "ReplicaMetadataResponse deserialization received unexpected error");
```

Comment:
After we set the exception, shall we set state = ReplicaGroupState.DONE?
Same in the following "else" condition.

Reply:
Never mind, the setException method has done that.